tools/tests_tools/test_generate_yaml_config.py

```diff
@@ -0,0 +1,241 @@
+
+    my_int: int = 42
+    """A positive integer value."""
+
+    my_float: float = 3.14
+    """A floating point value."""
+
+    my_str: str = 'foo'
+    """A string value."""
+
+    my_secret: ops.Secret | None = None
+    """A user secret."""
+
+    my_date: datetime.date = datetime.date(2023, 9, 4)
+    """A date value."""
+
+
+@dataclasses.dataclass(frozen=True, kw_only=True)
+class MyDataclassConfig:
+    basic_bool: bool
+    basic_int: int
+    basic_float: float
+    basic_str: str
+
+    my_bool: bool = False
+    """A Boolean value."""
+
+    my_int: int = 42
+    """A positive integer value."""
+
+    my_float: float = 3.14
+    """A floating point value."""
+
+    my_str: str = 'foo'
+    """A string value."""
+
+    my_secret: ops.Secret | None = None
+    """A user secret."""
+
+    my_date: datetime.date = datetime.date(2023, 9, 4)
+    """A date value."""
+
+
+_test_config_classes: list[type[object]] = [MyConfig, MyDataclassConfig]
+
+if pydantic:
+
+    @pydantic.dataclasses.dataclass(frozen=True, config={'arbitrary_types_allowed': True})
+    class MyPydanticDataclassConfig:
+        basic_bool: bool
+        basic_int: int
+        basic_float: float
+        basic_str: str
+        my_bool: bool = pydantic.Field(False, description='A Boolean value.')
+        my_int: int = pydantic.Field(42, description='A positive integer value.')
+        my_float: float = pydantic.Field(3.14, description='A floating point value.')
+        my_str: str = pydantic.Field('foo', description='A string value.')
+        my_secret: ops.Secret | None = pydantic.Field(None, description='A user secret.')
+        my_date: datetime.date = pydantic.Field(
+            datetime.date(2023, 9, 4), description='A date value.'
+        )
+
+    class MyPydanticBaseModelConfig(pydantic.BaseModel):
+        basic_bool: bool
+        basic_int: int
+        basic_float: float
+        basic_str: str
+        my_bool: bool = pydantic.Field(False, description='A Boolean value.')
+        my_int: int = pydantic.Field(42, description='A positive integer value.')
+        my_float: float = pydantic.Field(3.14, description='A floating point value.')
+        my_str: str = pydantic.Field('foo', description='A string value.')
+        my_secret: ops.Secret | None = pydantic.Field(None, description='A user secret.')
+        my_date: datetime.date = pydantic.Field(
+            datetime.date(2023, 9, 4), description='A date value.'
+        )
+
+        class Config:
+            arbitrary_types_allowed = True
+            frozen = True
+
+    _test_config_classes.extend((MyPydanticDataclassConfig, MyPydanticBaseModelConfig))
+
+
+@pytest.mark.parametrize('config_class', _test_config_classes)
+def test_config_yaml_schema(config_class: type[object]):
+    generated_schema = ops_tools.config_to_juju_schema(config_class)
+    expected_schema = {
+        'options': {
+            'basic-bool': {
+                'type': 'boolean',
+            },
```

Comment:
This is great, but omits YAML formatting.

There needs to be a test for verbatim YAML output.
I'm thinking a stability test perhaps (we may change the code, but the output keeps same "canonical" indent).